File: Chapter07/Chapter 07_Recomm_Engine.R


rm(list = ls())

# First change the following directory link to where all the input files do exist
setwd("D:\\Book writing\\Codes\\Chapter 7\\ml-latest-small\\ml-latest-small")

ratings = read.csv("ratings.csv")
movies = read.csv("movies.csv")

ratings = ratings[,!names(ratings) %in% c("timestamp")]

library(reshape2)

# Creating Pivot table
ratings_mat = acast(ratings,userId~movieId)
ratings_mat[is.na(ratings_mat)] =0


# Content based filtering
library(lsa)

a = c(2, 1, 0, 2, 0, 1, 1, 1)
b = c(2, 1, 1, 1, 1, 0, 1, 1)

print (paste("Cosine similarity between A and B is", round(cosine(a,b),4)))


m = nrow(ratings_mat);n = ncol(ratings_mat)


# User similarity matrix
mat_users = matrix(nrow = m, ncol = m)

for (i in 1:m){
  for (j in 1:m){
    if (i != j){
      mat_users[i,j] = cosine(ratings_mat[i,],ratings_mat[j,])
    }
    else {
      mat_users[i,j] = 0.0
    }
  }
}


colnames(mat_users) = rownames(ratings_mat); rownames(mat_users) = rownames(ratings_mat)
df_users = as.data.frame(mat_users)



# Finiding similar users
topn_simusers <- function(uid=16,n=5){
  sorted_df = sort(df_users[uid,],decreasing = TRUE)[1:n]
  print(paste("Similar users as user:",uid))
  return(sorted_df)
}

print(topn_simusers(uid = 17,n=10))


# Finiding most rated movies of a user
library(sqldf)

ratings_withmovie = sqldf(" select a.*,b.title from 
                          ratings as a left join movies as b 
                          on a.movieId = b.movieId")


# Finding most rated movies of a user
topn_movieratings <- function(uid=355,n_ratings=10){
  uid_ratings = ratings_withmovie[ratings_withmovie$userId==uid,]
  sorted_uidrtng = uid_ratings[order(-uid_ratings$rating),]
  return(head(sorted_uidrtng,n_ratings))
}
  
print( topn_movieratings(uid = 596,n=10))



# Movies similarity matrix
mat_movies = matrix(nrow = n, ncol = n)

for (i in 1:n){
  for (j in 1:n){
    if (i != j){
      mat_movies[i,j] = cosine(ratings_mat[,i],ratings_mat[,j])
    }
    else {
      mat_movies[i,j] = 0.0
    }
  }
}

colnames(mat_movies) = colnames(ratings_mat); rownames(mat_movies) = colnames(ratings_mat)
df_movies = as.data.frame(mat_movies)

write.csv(df_movies,"df_movies.csv")

df_movies = read.csv("df_movies.csv")
rownames(df_movies) = df_movies$X
colnames(df_movies) = c("aaa",df_movies$X)
df_movies = subset(df_movies, select=-c(aaa))



# Finiding similar movies
topn_simovies <- function(mid=588,n_movies=5){
  sorted_df = sort(df_movies[mid,],decreasing = TRUE)[1:n_movies]
  sorted_df_t = as.data.frame(t(sorted_df))
  colnames(sorted_df_t) = c("score")
  sorted_df_t$movieId = rownames(sorted_df_t)
  
  print(paste("Similar",n_movies, "movies as compared to the movie",mid,"are :"))
  sorted_df_t_wmovie = sqldf(" select a.*,b.title from sorted_df_t as a left join movies as b 
                          on a.movieId = b.movieId")
  return(sorted_df_t_wmovie)
}

print(topn_simovies(mid = 589,n_movies=15))




# Collaborative filtering
ratings = read.csv("ratings.csv")
movies = read.csv("movies.csv")

library(sqldf)
library(reshape2)
library(recommenderlab)

ratings_v2 = ratings[,-c(4)]

ratings_mat = acast(ratings_v2,userId~movieId)
ratings_mat2 =  as(ratings_mat, "realRatingMatrix")

getRatingMatrix(ratings_mat2)

#Plotting user-item complete matrix
image(ratings_mat2, main = "Raw Ratings")

# Fitting ALS method on Data
rec=Recommender(ratings_mat2[1:nrow(ratings_mat2)],method="UBCF", param=list(normalize = "Z-score",method="Cosine",nn=5, minRating=1))
rec_2=Recommender(ratings_mat2[1:nrow(ratings_mat2)],method="POPULAR")

print(rec)
print(rec_2)

names(getModel(rec))
getModel(rec)$nn


# Create predictions for all the users
recom_pred = predict(rec,ratings_mat2[1:nrow(ratings_mat2)],type="ratings")

# Putting predicitons into list
rec_list<-as(recom_pred,"list")
head(summary(rec_list))

print_recommendations <- function(uid=586,top_nmovies=10){
  recoms_list = rec_list[[uid]]
  sorted_df = as.data.frame(sort(recoms_list,decreasing = TRUE)[1:top_nmovies])
  colnames(sorted_df) = c("score")

  sorted_df$movieId = rownames(sorted_df)
  print(paste("Movies recommended for the user",uid,"are follows:"))
  sorted_df_t_wmovie = sqldf(" select a.*,b.title from sorted_df as a left join movies as b 
                             on a.movieId = b.movieId")
  
  return(sorted_df_t_wmovie)
}

print(print_recommendations(uid = 580,top_nmovies = 15))

File: Chapter07/Chapter 07_Recomm_Engine.py


import os
""" First change the following directory link to where all input files do exist """
os.chdir("D:\\Book writing\\Codes\\Chapter 7\\ml-latest-small\\ml-latest-small")


import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


ratings = pd.read_csv("ratings.csv")
print (ratings.head())

movies = pd.read_csv("movies.csv")
print (movies.head())


#Combining movie ratings & movie names
ratings = pd.merge(ratings[['userId','movieId','rating']],movies[['movieId','title']],
                   how='left',left_on ='movieId' ,right_on = 'movieId')


rp = ratings.pivot_table(columns = ['movieId'],index = ['userId'],values = 'rating')
rp = rp.fillna(0)

# Converting pandas dataframe to numpy for faster execution in loops etc.
rp_mat = rp.as_matrix()


from scipy.spatial.distance import cosine


#The cosine of the angle between them is about 0.822.
a= np.asarray( [2, 1, 0, 2, 0, 1, 1, 1])
b = np.asarray( [2, 1, 1, 1, 1, 0, 1, 1])

print("\n\n")
print ("Cosine similarity between A and B is",round(1-cosine(a,b),4))


m, n = rp.shape

# User similarity matrix
mat_users = np.zeros((m, m))
for i in range(m):
    for j in range(m):
        if i != j:
            mat_users[i][j] = (1- cosine(rp_mat[i,:], rp_mat[j,:]))
        else:
            mat_users[i][j] = 0.
            
pd_users = pd.DataFrame(mat_users,index =rp.index ,columns= rp.index )


# Finding similar users
def topn_simusers(uid = 16,n=5):
    users = pd_users.loc[uid,:].sort_values(ascending = False)
    topn_users = users.iloc[:n,]
    topn_users = topn_users.rename('score')    
    print ("Similar users as user:",uid)
    return pd.DataFrame(topn_users)

print (topn_simusers(uid=17,n=10))   


# Finding most rated movies of a user
def topn_movieratings(uid = 355,n_ratings=10):    
    uid_ratings = ratings.loc[ratings['userId']==uid]
    uid_ratings = uid_ratings.sort_values(by='rating',ascending = [False])
    print ("Top",n_ratings ,"movie ratings of user:",uid)
    return uid_ratings.iloc[:n_ratings,]    

print (topn_movieratings(uid=596,n_ratings=10))


# Movie similarity matrix
import time
start_time = time.time()
mat_movies = np.zeros((n, n))

for i in range(n):
    for j in range(n):
        if i!=j:
            mat_movies[i,j] = (1- cosine(rp_mat[:,i], rp_mat[:,j]))
        else:
            mat_movies[i,j] = 0.
print("--- %s seconds ---" % (time.time() - start_time))


pd_movies = pd.DataFrame(mat_movies,index =rp.columns ,columns= rp.columns )


#pd_movies.to_csv('pd_movies.csv',sep=',')
pd_movies = pd.read_csv("pd_movies.csv",index_col='movieId')


# Finding similar movies
def topn_simovies(mid = 588,n=15):
    mid_ratings = pd_movies.loc[mid,:].sort_values(ascending = False)
    topn_movies = pd.DataFrame(mid_ratings.iloc[:n,])
    topn_movies['index1'] = topn_movies.index
    topn_movies['index1'] = topn_movies['index1'].astype('int64')
    topn_movies = pd.merge(topn_movies,movies[['movieId','title']],how = 'left',left_on ='index1' ,right_on = 'movieId')
    print ("Movies similar to movie id:",mid,",",movies['title'][movies['movieId']==mid].to_string(index=False),",are")
    del topn_movies['index1']
    return topn_movies


print (topn_simovies(mid=589,n=15))




#Collaborative filtering

import os
""" First change the following directory link to where all input files do exist """
os.chdir("D:\\Book writing\\Codes\\Chapter 7\\ml-latest-small\\ml-latest-small")


import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


ratings = pd.read_csv("ratings.csv")
print (ratings.head())

movies = pd.read_csv("movies.csv")
print (movies.head())

rp = ratings.pivot_table(columns = ['movieId'],index = ['userId'],values = 'rating')
rp = rp.fillna(0)

A = rp.values

print ("\nShape of Original Sparse Matrix",A.shape)


W = A>0.5
W[W==True]=1
W[W==False]=0
W = W.astype(np.float64,copy=False)


W_pred = A<0.5
W_pred[W_pred==True]=1
W_pred[W_pred==False]=0
W_pred = W_pred.astype(np.float64,copy=False)
np.fill_diagonal(W_pred,val=0)



# Parameters
m,n = A.shape

n_iterations = 200
n_factors = 100
lmbda = 0.1

X = 5 * np.random.rand(m,n_factors)
Y = 5* np.random.rand(n_factors,n)

def get_error(A, X, Y, W):
    return np.sqrt(np.sum((W * (A - np.dot(X, Y)))**2)/np.sum(W))

errors = []
for itr in range(n_iterations):
    X = np.linalg.solve(np.dot(Y,Y.T)+ lmbda * np.eye(n_factors),np.dot(Y,A.T)).T
    Y = np.linalg.solve(np.dot(X.T,X)+ lmbda * np.eye(n_factors),np.dot(X.T,A))
    
    if itr%10 == 0:
        print(itr," iterations completed","RMSError value is:",get_error(A,X,Y,W))
                 
    errors.append(get_error(A,X,Y,W))

A_hat = np.dot(X,Y)
print ("RMSError of rated movies: ",get_error(A,X,Y,W))
    

plt.plot(errors);
plt.ylim([0, 3.5]);
plt.xlabel("Number of Iterations");plt.ylabel("RMSE")
#plt.title("No.of Iterations vs. RMSE")
plt.show()



def print_recommovies(uid=315,n_movies=15,pred_mat = A_hat,wpred_mat = W_pred ):
    pred_recos = pred_mat*wpred_mat
    pd_predrecos = pd.DataFrame(pred_recos,index =rp.index ,columns= rp.columns )
    pred_ratings = pd_predrecos.loc[uid,:].sort_values(ascending = False)
    pred_topratings = pred_ratings[:n_movies,]
    pred_topratings = pred_topratings.rename('pred_ratings')  
    pred_topratings = pd.DataFrame(pred_topratings)
    pred_topratings['index1'] = pred_topratings.index
    pred_topratings['index1'] = pred_topratings['index1'].astype('int64')
    pred_topratings = pd.merge(pred_topratings,movies[['movieId','title']],how = 'left',left_on ='index1' ,right_on = 'movieId')
    del pred_topratings['index1']    
    print ("\nTop",n_movies,"movies predicted for the user:",uid," based on collaborative filtering\n")
    return pred_topratings


predmtrx = print_recommovies(uid=355,n_movies=10,pred_mat=A_hat,wpred_mat=W_pred)
print (predmtrx)




# Grid Search on Collaborative Filtering
def get_error(A, X, Y, W):
    return np.sqrt(np.sum((W * (A - np.dot(X, Y)))**2)/np.sum(W))

niters = [20,50,100,200]
factors = [30,50,70,100]
lambdas = [0.001,0.01,0.05,0.1]

init_error = float("inf")


print("\n\nGrid Search results of ALS Matrix Factorization:\n")
for niter in niters:
    for facts in factors:
        for lmbd in lambdas:
                
            X = 5 * np.random.rand(m,facts)
            Y = 5* np.random.rand(facts,n)
            
            for itr in range(niter):
                X = np.linalg.solve(np.dot(Y,Y.T)+ lmbd * np.eye(facts),np.dot(Y,A.T)).T
                Y = np.linalg.solve(np.dot(X.T,X)+ lmbd * np.eye(facts),np.dot(X.T,A))
            
            error = get_error(A,X,Y,W)
            
            if error<init_error:
                print ("No.of iters",niter,"No.of Factors",facts,"Lambda",lmbd,"RMSE",error)
                init_error = error
                
            
            












